### ISSUE TITLE
Accessing Internal User map

### ISSUE THREAD

**Issue by djshadowxm82** [uhppoted/uhppote-core#5]
So is there away to acquire the map of the user's id card, to the name you set in the web interface? Is that accessible anywhere aside from me writing a ridiculous web parser to access it from the table structure?

**Comment by uhppoted** [comment]
Ooops .. was *not* aware that anybody other than me had even looked at the (highly experimental) web interface - it's still very much _in development_.

The controllers themselves do not store anything else than the card number so the user name is always mapped from an external database of some kind (spreadsheet, file, API, etc)

Where exactly do you want to lookup the user name i.e. in the browser, on the HTTP server, using the CLI, etc? 
- For the CLI you would have to use something like _jq_ or _sqlite3_.
- On the HTTP server it's all stored internally but also mirrored to a JSON file (e.g. _/var/uhppoted/httpd/system/cards.json_ on Linux)
- In the browser, the information is stored locally in [db.js](https://github.com/uhppoted/uhppoted-httpd/blob/master/httpd/html/javascript/db.js)

If you can give me a bit more detail about what you want to do?

**Comment by djshadowxm82** [comment]
Let me rephrase, I was talking from the controller itself. When you directly add cards to the firmware based system, it allows you to add a name to the card and it stays with the system. I have not found away to read that information back.

**Comment by uhppoted** [comment]
There are two APIs:
- the internal one used by the manufacturer software which is private and not made available
- the public API which is the one used by `uhppote-core`

I'm guessing that the internal API has additional functionality that includes the name assigned to the card number - the public API doesn't have any way to set/retrieve it.

**Comment by djshadowxm82** [comment]
> There are two APIs:
> - the internal one used by the manufacturer software which is private and not made available
> - the public API which is the one used by `uhppote-core`
> 
> I'm guessing that the internal API has additional functionality that includes the name assigned to the card number - the public API doesn't have any way to set/retrieve it (or at least not in the documentation I have).

I was hoping you maybe had stumbled upon a method to retrieve this information. I've been digging for a few days in the c# disassembly trying to see if the Chinese software had mention of it and I cannot seem to find it there either. It is infact being stored somewhere because the information persists after power loss.

**Comment by uhppoted** [comment]
Ah,ok .. I'm afraid I don't have anything for you :-(. 

You might be able to pick it out of the network protocol using Wireshark, but TBH I haven't had much success with reverse engineering it (admittedly I didn't try very hard).

**Comment by djshadowxm82** [comment]
> Ah,ok .. I'm afraid I don't have anything for you :-(. 
> 
> You might be able to pick it out of the network protocol using Wireshark, but TBH I haven't had much success with reverse engineering it (admittedly I didn't try very hard).
> 
> 

Dnspy. Tons of stuff in those c# apps. I'm working on. Interface for customer thats more user friendly than both web and c# app. I just wad hoping I didn't have to have a running db somewhere because I need it to be fairly minimalistic.

**Comment by uhppoted** [comment]
Is there any possibility the controllers are phoning home to e.g. a cloud DB ? The original controller UI used an MS Access DB to store user information so I've always just assumed it wasn't being stored on the controllers.

re. customer interface. I do have a [DLL](https://github.com/uhppoted/uhppoted-dll) with a [C# binding](https://github.com/uhppoted/uhppoted-dll/tree/master/bindings/csharp) if that's of any use to you. But yeah, it won't include anything that's not in the public API.

**Comment by djshadowxm82** [comment]
> Is there any possibility the controllers are phoning home to e.g. a cloud DB ? The original controller UI used an MS Access DB to store user information so I've always just assumed it wasn't being stored on the controllers.
> 
> re. customer interface. I do have a [DLL](https://github.com/uhppoted/uhppoted-dll) with a [C# binding](https://github.com/uhppoted/uhppoted-dll/tree/master/bindings/csharp) if that's of any use to you. But yeah, it won't include anything that's not in the public API.
> 
> 

No, it's definitely in the flash somewhere, because when you wipe the controller with a special packet from the software. The users lists is gone. Plus the information isn't syncing to the software.

**Comment by uhppoted** [comment]
Hmm .. interesting! That would be really useful .. there are some unused bytes in the official message structure and maybe they can be used for ad hoc stuff. It's a bit of a reach and I'm not honestly very hopeful but it's worth a try - I'll play a bit tomorrow.

**Comment by uhppoted** [comment]
Having a dump of the firmware would be incredibly useful! There was somebody else who was going to try extract it but I haven't heard from him in a couple of months - I'll ping him and see how far he's gotten.

However - I modified my `get/put-card` functions to fill the _unused_ bytes in the request message and it turns out they're not _unused_, just entirely undocumented so maybe the name information may be coming back. Can you try this:

- Get the [uhppote-cli](https://github.com/uhppoted/uhppote-cli) for your platform (there are executables in the [Releases](https://github.com/uhppoted/uhppote-cli/releases) page if you don't want to build it yourself)
- Run this command:
```
uhppote-cli --debug get-card <controller serial number> <card number>
```
e.g.
```
uhppote-cli --debug get-card 405419896 8165538
```
The bytes from 24 to 64 are the undocumented fields and possibly the name might be in there somewhere.

**Comment by djshadowxm82** [comment]
![image](https://user-images.githubusercontent.com/10658965/169152924-3681819e-7c5f-493e-a795-0bbf7e59bd8e.png)


> Having a dump of the firmware would be incredibly useful! There was somebody else who was going to try extract it but I haven't heard from him in a couple of months - I'll ping him and see how far he's gotten.
> 
> However - I modified my `get/put-card` functions to fill the _unused_ bytes in the request message and it turns out they're not _unused_, just entirely undocumented so maybe the name information may be coming back. Can you try this:
> 
> * Get the [uhppote-cli](https://github.com/uhppoted/uhppote-cli) for your platform (there are executables in the [Releases](https://github.com/uhppoted/uhppote-cli/releases) page if you don't want to build it yourself)
> * Run this command:
> 
> ```
> uhppote-cli --debug get-card <controller serial number> <card number>
> ```
> 
> e.g.
> 
> ```
> uhppote-cli --debug get-card 405419896 8165538
> ```
> 
> The bytes from 24 to 64 are the undocumented fields and possibly the name might be in there somewhere.

![image](https://user-images.githubusercontent.com/10658965/169152933-9e5765ba-d44b-4114-ac71-28e463ba489d.png)

**Comment by djshadowxm82** [comment]
> > Having a dump of the firmware would be incredibly useful! There was somebody else who was going to try extract it but I haven't heard from him in a couple of months - I'll ping him and see how far he's gotten.
> > However - I modified my `get/put-card` functions to fill the _unused_ bytes in the request message and it turns out they're not _unused_, just entirely undocumented so maybe the name information may be coming back. Can you try this:
> > 
> > * Get the [uhppote-cli](https://github.com/uhppoted/uhppote-cli) for your platform (there are executables in the [Releases](https://github.com/uhppoted/uhppote-cli/releases) page if you don't want to build it yourself)
> > * Run this command:
> > 
> > ```
> > uhppote-cli --debug get-card <controller serial number> <card number>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > e.g.
> > ```
> > uhppote-cli --debug get-card 405419896 8165538
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The bytes from 24 to 64 are the undocumented fields and possibly the name might be in there somewhere.
> 
> ![image](https://user-images.githubusercontent.com/10658965/169152933-9e5765ba-d44b-4114-ac71-28e463ba489d.png)

![image](https://user-images.githubusercontent.com/10658965/169153074-5aed3586-ff53-4b3a-a89d-30ddf2555a3e.png)

**Comment by uhppoted** [comment]
Hmm, ok I guess not :-(. I was able to store and retrieve bytes 24-32 so those might be just a scratchpad.

**Comment by djshadowxm82** [comment]
> Hmm, ok I guess not :-(. I was able to store and retrieve bytes 24-32 so those might be just a scratchpad.
> 
> Oh, wait .. digging through the documentation it seems:
> 
> * bytes 24-26 are the user password (PIN code)
> * bytes 27-31 are the first card/more card privileges
> * bytes 40-43 are the sequence number

![image](https://user-images.githubusercontent.com/10658965/169343107-89d69e24-e28b-4266-9e95-1cad32bda108.png)

Interesting. Well I dumped the memory chip this morning and un interesting results were found. 

This was at memory address 0x00001000 and the rest seemed to be uninitialized.  I was hopeful that maybe that chip which has 8MB storage to it had something worth while on it, but its possible that everything else is stored elsewhere. Possibly on internal memory in the NXP chip which is above my paygrade so to speak in accessing information off of. Especially on a board I need to put into a production environment. But this is something I supposed. It's possible that this flash memory is for long term storage?

**Comment by uhppoted** [comment]
It's a start at least :-) ... I've pinged Chris to see if he's had any luck with getting a dump of the firmware.

re. 8MB flash. Best guess is that it's probably for storing the event log - the controllers can store about 200 000 events on board (or at least mine can - newer boards may have more storage) and 16 bytes may be just enough for one event if it's byte-packed. 

 I'd be curious to know if the flash changes after e.g. swiping a card.

But - 200 000 events @16bytes each is under 4MB so there's loads of room for other stuff. Can't remember where I saw it but I think the boards can store up to 20000 cards? I need to confirm that, but if that is the case then I would expect them to be in the flash too because the LPC1766FBD100 only has 512kb on-board flash which isn't nearly enough.

**Comment by djshadowxm82** [comment]
Agreed, let me know your findings.

**Comment by djshadowxm82** [comment]
Well Might have found something interesting for you to play with @uhppoted. 

`internal int DownloadIP_internal(int ControllerSN, string IP, int Port, string doorName, ref DataTable dtPrivilege, string PCIPAddr)
		{
			if (wgMjControllerPrivilege.m_bStopDownloadPrivilege)
			{
				return -1;
			}
			bool flag = false;
			bool flag2 = false;
			bool flag3 = false;
			bool flag4 = false;
			if (dtPrivilege.Columns.Contains("f_DoorFirstCard_1"))
			{
				flag = true;
			}
			if (dtPrivilege.Columns.Contains("f_MoreCards_GrpID_1"))
			{
				flag2 = true;
			}
			if (dtPrivilege.Columns.Contains("f_IsDeleted"))
			{
				flag3 = true;
			}
			else
			{
				dtPrivilege.Columns.Add("f_IsDeleted", Type.GetType("System.UInt32"));
			}
			if (dtPrivilege.Columns.Contains("f_ConsumerName"))
			{
				flag4 = true;
			}
			wgTools.WriteLine("Download Privilege Start");
			WGPacket wgpacket = new WGPacket();
			wgpacket.type = 35;
			wgpacket.code = 16;
			wgpacket.iDevSnFrom = 0u;
			wgpacket.iDevSnTo = (uint)ControllerSN;
			wgpacket.iCallReturn = 0;
			wgTools.getUdpComm(ref this.wgudp, -1);
			byte[] array = wgpacket.ToBytes(this.wgudp.udpPort);
			if (array == null)
			{
				return -1;
			}
			byte[] array2 = null;
			int num = this.wgudp.udp_get(array, 300, wgpacket.xid, IP, Port, ref array2);
			MjRegisterCardsParam mjRegisterCardsParam = new MjRegisterCardsParam();
			if (array2 != null)
			{
				mjRegisterCardsParam.updateParam(array2, 20);
				WGPacketSSI_FLASH_QUERY_internal wgpacketSSI_FLASH_QUERY_internal = new WGPacketSSI_FLASH_QUERY_internal();
				wgpacketSSI_FLASH_QUERY_internal = new WGPacketSSI_FLASH_QUERY_internal(33, 16, (uint)ControllerSN, 0u, 1023u);
				long num2 = 0L;
				while (num2 < (long)((ulong)(mjRegisterCardsParam.totalPrivilegeCount - mjRegisterCardsParam.deletedPrivilegeCount)))
				{
					long num3;
					if ((ulong)((mjRegisterCardsParam.newPrivilegePage4KAddr - mjRegisterCardsParam.iPrivilegeFirstIndexSSI_FLASH_PRIVILEGE_STARTADDR) / 8u) > (ulong)num2)
					{
						num3 = (long)((ulong)mjRegisterCardsParam.iPrivilegeFirstIndexSSI_FLASH_PRIVILEGE_STARTADDR + (ulong)(num2 * 8L));
					}
					else
					{
						num3 = (long)((ulong)(mjRegisterCardsParam.newPrivilegePage4KAddr + 4096u) + (ulong)((num2 - (long)((ulong)((mjRegisterCardsParam.newPrivilegePage4KAddr - mjRegisterCardsParam.iPrivilegeFirstIndexSSI_FLASH_PRIVILEGE_STARTADDR) / 8u))) * 8L));
					}
					wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr = (uint)num3;
					wgpacketSSI_FLASH_QUERY_internal.iEndFlashAddr = (uint)(num3 + 1024L - 1L);
					byte[] array3 = wgpacketSSI_FLASH_QUERY_internal.ToBytes(this.wgudp.udpPort);
					if (array3 == null)
					{
						return -1;
					}
					byte[] array4 = null;
					num = this.wgudp.udp_get(array3, 300, wgpacketSSI_FLASH_QUERY_internal.xid, IP, Port, ref array4);
					if (num < 0)
					{
						return -1;
					}
					MjRegisterCard mjRegisterCard = new MjRegisterCard();
					byte[] array5 = new byte[24];
					if (array4 != null)
					{
						Array.Copy(array4, 28, array5, 0, 8);
					}
					num3 = num3 * 2L + 1654784L;
					wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr = (uint)num3;
					wgpacketSSI_FLASH_QUERY_internal.iEndFlashAddr = (uint)(num3 + 1024L - 1L);
					array3 = wgpacketSSI_FLASH_QUERY_internal.ToBytes(this.wgudp.udpPort);
					if (array3 == null)
					{
						return -1;
					}
					byte[] array6 = null;
					num = this.wgudp.udp_get(array3, 300, wgpacketSSI_FLASH_QUERY_internal.xid, IP, Port, ref array6);
					if (num < 0)
					{
						return -1;
					}
					if (array6 != null)
					{
						for (int i = 0; i < 64; i++)
						{
							Array.Copy(array4, 28 + i * 8, array5, 0, 8);
							Array.Copy(array6, 28 + i * 16, array5, 8, 16);
							int count = dtPrivilege.Rows.Count;
							if (mjRegisterCard.Update(array5, (uint)(num2 + (long)i)) <= 0)
							{
								break;
							}
							if (wgMjControllerPrivilege.bStopDownloadPrivilege)
							{
								return -100002;
							}
							DataRow dataRow = dtPrivilege.NewRow();
							dataRow["f_CardNO"] = mjRegisterCard.CardID;
							dataRow["f_PIN"] = mjRegisterCard.Password.ToString();
							dataRow["f_BeginYMD"] = mjRegisterCard.ymdStart;
							dataRow["f_EndYMD"] = mjRegisterCard.ymdEnd;
							dataRow["f_ControlSegID1"] = mjRegisterCard.ControlSegIndexGet(1);
							dataRow["f_ControlSegID2"] = mjRegisterCard.ControlSegIndexGet(2);
							dataRow["f_ControlSegID3"] = mjRegisterCard.ControlSegIndexGet(3);
							dataRow["f_ControlSegID4"] = mjRegisterCard.ControlSegIndexGet(4);
							if (flag)
							{
								dataRow["f_DoorFirstCard_1"] = mjRegisterCard.FirstCardGet(1);
								dataRow["f_DoorFirstCard_2"] = mjRegisterCard.FirstCardGet(2);
								dataRow["f_DoorFirstCard_3"] = mjRegisterCard.FirstCardGet(3);
								dataRow["f_DoorFirstCard_4"] = mjRegisterCard.FirstCardGet(4);
							}
							if (flag2)
							{
								dataRow["f_MoreCards_GrpID_1"] = mjRegisterCard.MoreCardGroupIndexGet(1);
								dataRow["f_MoreCards_GrpID_2"] = mjRegisterCard.MoreCardGroupIndexGet(2);
								dataRow["f_MoreCards_GrpID_3"] = mjRegisterCard.MoreCardGroupIndexGet(3);
								dataRow["f_MoreCards_GrpID_4"] = mjRegisterCard.MoreCardGroupIndexGet(4);
							}
							dataRow["f_IsDeleted"] = (mjRegisterCard.IsDeleted ? 1 : 0);
							dataRow["f_AllowFloors"] = mjRegisterCard.AllowFloors;
							dtPrivilege.Rows.Add(dataRow);
						}
						dtPrivilege.AcceptChanges();
					}
					num2 += 64L;
					if (wgMjControllerPrivilege.m_bStopDownloadPrivilege)
					{
						return -1;
					}
				}
				if ((ulong)mjRegisterCardsParam.totalPrivilegeCount > (ulong)((long)this.m_uploadUserNameMaxlen))
				{
					flag4 = false;
				}
				if (flag4)
				{
					int userNamePageStart = this.m_userNamePageStart;
					wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr = (uint)(userNamePageStart * 1024);
					wgpacketSSI_FLASH_QUERY_internal.iEndFlashAddr = wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr + 1024u - 1u;
					if (mjRegisterCardsParam.iPrivilegeFirstIndexSSI_FLASH_PRIVILEGE_STARTADDR > 0u)
					{
						wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr = (uint)(this.m_userNamePageHalfStart * 1024);
					}
					int num4 = 0;
					while (wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr <= 4956159u)
					{
						num = this.wgudp.udp_get(wgpacketSSI_FLASH_QUERY_internal.ToBytes(this.wgudp.udpPort), 300, wgpacketSSI_FLASH_QUERY_internal.xid, IP, Port, ref array2);
						if (num < 0)
						{
							return -1;
						}
						if (array2 != null)
						{
							for (int j = 0; j < 1024; j += 32)
							{
								byte[] array7 = new byte[33];
								int num5 = 0;
								array7[32] = 0;
								for (int k = 0; k < 32; k++)
								{
									array7[k] = array2[28 + j + k];
									if (array7[k] == 0)
									{
										break;
									}
									num5++;
								}
								if ((array7[0] != 255 || array7[1] != 255) && array7[0] != 0)
								{
									if (num4 >= dtPrivilege.Rows.Count)
									{
										break;
									}
									string @string = Encoding.GetEncoding("utf-8").GetString(array7, 0, num5);
									dtPrivilege.Rows[num4]["f_ConsumerName"] = @string;
								}
								num4++;
							}
						}
						else
						{
							num4 += 32;
						}
						if (num4 >= dtPrivilege.Rows.Count)
						{
							IL_709:
							dtPrivilege.AcceptChanges();
							goto IL_711;
						}
						wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr += 1024u;
					}
					goto IL_709;
				}
				IL_711:
				DataView dataView = new DataView(dtPrivilege);
				dataView.RowFilter = "f_IsDeleted = 1";
				if (dataView.Count > 0)
				{
					for (int l = dataView.Count - 1; l >= 0; l--)
					{
						dataView.Delete(l);
					}
				}
				dtPrivilege.AcceptChanges();
				if (!flag3)
				{
					dtPrivilege.Columns.Remove("f_IsDeleted");
					dtPrivilege.AcceptChanges();
				}
				return 1;
			}
			return -1;
		}`


Found this snippet of code within the c# specifically 

this section. flag4 aka consumername (which i think is the card name in flash)

``	p
`if (flag4)
				{
                                        private readonly int m_userNamePageStart = 4600;
                                     	private readonly int m_userNamePageHalfStart = 4720;

					int userNamePageStart = this.m_userNamePageStart;
					wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr = (uint)(userNamePageStart * 1024);
					wgpacketSSI_FLASH_QUERY_internal.iEndFlashAddr = wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr + 1024u - 1u;
					if (mjRegisterCardsParam.iPrivilegeFirstIndexSSI_FLASH_PRIVILEGE_STARTADDR > 0u)
					{
						wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr = (uint)(this.m_userNamePageHalfStart * 1024);
					}
					int num4 = 0;
					while (wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr <= 4956159u)
					{
						num = this.wgudp.udp_get(wgpacketSSI_FLASH_QUERY_internal.ToBytes(this.wgudp.udpPort), 300, wgpacketSSI_FLASH_QUERY_internal.xid, IP, Port, ref array2);
						if (num < 0)
						{
							return -1;
						}
						if (array2 != null)
						{
							for (int j = 0; j < 1024; j += 32)
							{
								byte[] array7 = new byte[33];
								int num5 = 0;
								array7[32] = 0;
								for (int k = 0; k < 32; k++)
								{
									array7[k] = array2[28 + j + k];
									if (array7[k] == 0)
									{
										break;
									}
									num5++;
								}
								if ((array7[0] != 255 || array7[1] != 255) && array7[0] != 0)
								{
									if (num4 >= dtPrivilege.Rows.Count)
									{
										break;
									}
									string @string = Encoding.GetEncoding("utf-8").GetString(array7, 0, num5);
									dtPrivilege.Rows[num4]["f_ConsumerName"] = @string;
								}
								num4++;
							}
						}
						else
						{
							num4 += 32;
						}
						if (num4 >= dtPrivilege.Rows.Count)
						{
							IL_709:
							dtPrivilege.AcceptChanges();
							goto IL_711;
						}
						wgpacketSSI_FLASH_QUERY_internal.iStartFlashAddr += 1024u;
					}
					goto IL_709;
				}`

**Comment by djshadowxm82** [comment]
Let me know if you find any of that useful!

**Comment by uhppoted** [comment]
Whoa!!! That is **interesting** .. nice find!!! So it looks like it may be possible to read (some of) the flash remotely .. they need to pay their software engineers better :-) :-).

I've seen references to `n3k_comm.dll` in the old SDK samples - AFAIK it's used to encrypt/decrypt the UDP messages (using the password as the key) which is why reverse engineering the protocol from Wireshark traces is not entirely straightforward. Hopefully the encryption is something standard or else not too convoluted if it's home-grown.

'k, I have some dumb questions because I'm on a Mac (and/or Linux) and have never been able to run the Windows software (original reason for this project actually):
- the code above is disassembled from the manufacturer's Windows application?
- what version are you using? I'll have to go looking for it because I only have an ancient version from 4-5 years back..

**Comment by uhppoted** [comment]
BTW, Chris got back to me - he hasn't been able to get a look at the firmware either and in the process of trying managed to brick a couple of his controllers (although that might have been a faulty programmer).

**Comment by djshadowxm82** [comment]
> Whoa!!! That is **interesting** .. nice find!!! So it looks like it may be possible to read (some of) the flash remotely.
> 
> I've seen references to `n3k_comm.dll` in the old SDK samples - AFAIK it's used to encrypt/decrypt the UDP messages (using the password as the key) which is why reverse engineering the protocol from Wireshark traces is not entirely straightforward. Hopefully the encryption is something standard or else not too convoluted if it's home-grown.
> 
> 'k, I have some dumb questions because I'm on a Mac/Linux and have never been able to run the Windows software (original reason for this project actually):
> - the code above is disassembled from the manufacturer's Windows application?
> - what version are you using? I'll have to go looking for it because I only have an ancient version from 4-5 years back..
> 
> 
> 
> 

It's all from the newest ( I used disassembly to find their server with newest.)

Its all from disassembly. Message me directly on Twitter. I'll get you what you need.

**Comment by uhppoted** [comment]
<smile> @uhppoted  is now following you on twitter :-) (can't DM you for reasons known only to twitter). 

'k, now to figure out how to tunnel broadcast UDP out of a Windows VPS...
